docs: update release process and add oCIS Web update steps (#12242)
owncloud/web: docs/releasing.md

### Release Guide

1. Checkout a stable branch
   * If you are releasing a new major version, create a new branch `stable-$major` and push it
   * If you are releasing a new minor or patch version, checkout the latest `stable-$major`

>    * If you are releasing a backport of a bugfix, checkout latest `stable-$major.$minor.$patch` (or `stable-$major` if patch branch is not available), create a new branch `stable-$major.$minor.$patch` and push it

Reviewer: Reading this, I am not fully clear about it. Just some notes:
To be added  in the line after item 1 but before the bullet item: Replace the placeholders `$major`, `$minor` or `$patch` accordingly
minor/patch: ...checkout the latest `stable-$major` --> is this really `stable-$major` and if yes we should note it why. Need to add `and push it`
Backport also needs some love to make it more clear

2. Create a new branch `chore/bump-version-$major.$minor.$patch`
3. Backport all required changes from `master` to the stable branch
4. Create a folder in `changelog` for the release version and date `mkdir $major.$minor.$patchVersion_YYYY-MM-DD`
5. Move all changelog items from the `changelog/unreleased/` folder to the `$major.$minor.$patchVersion_YYYY-MM-DD` folder
5. Pull translations `make l10n-pull`

> 6. Write translations `make l10n-write`

Reviewer: Where is this data going to, master or the target branch?

7. Run `./dev/scripts/bump_versions.sh $version`. This script will bump the `package.json` files in all relevant packages and the sonar cloud project version
8. Commit and push your changes.

> 9. Create a PR with tag `[full-ci]` in the title to merge the `chore/bump-version-$major.$minor.$patch` branch into the stable branch from step 1 (requires at least two approvals)

Reviewer: ```suggestion
9. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-version-$major.$minor.$patch` branch into the stable branch from step 1 (requires at least two approvals)
```

10. After merging, checkout the stable branch from step 1 again
11. Run `./dev/scripts/create_and_push_tags.sh`. This script will create and push tags for the main app as well as all packages that need to be released
12. The GitHub release will be created automatically together with the release artifacts

### Update Web in oCIS

> 1. checkout master branch

Reviewer: In which repo. Let's add this info as onw single line above.

2. create a new branch `chore/bump-web-to-$major.$minor.$patch`

> 3. update `WEB_COMMITID` in `.drone.env` to the commit id of the Web release tag
> 4. update `WEB_BRANCH` in `.drone.env` to the Web release stable branch
> 5. update `WEB_ASSETS_VERSION` in `services/web/Makefile` to the Web release version

Reviewer: ```suggestion
3. in `.drone.env`, update `WEB_COMMITID` to the commit id of the Web release tag
4. in `.drone.env`, update `WEB_BRANCH` to the Web release stable branch
5. in `services/web/Makefile`, update `WEB_ASSETS_VERSION` to the Web release version
```
Easier to see "where" first and then "what"

### Release Guide

> #### Releasing new major version

Reviewer: ```suggestion
#### Releasing New Major Version
```

> 1. Checkout the latest `stable-$major` branch \
> This branch will be used as the base for the new major version branch. \
> Replace the `$major` with the last released major version.
> 2. Create a new branch `stable-$major`. \
> Replace the `$major` with the new major version.
> 3. Push the newly created `stable-$major` branch
> 4. Create a new branch `chore/bump-version-$major.$minor.$patch` using the branch from previous step as base \
> Replace the `$major`, `$minor` and `$patch` with the new version.
> 5. Cherry pick all required changes from `master` into the current branch
> 6. Create a new folder in `/changelog/` folder using the release version and current date `mkdir $major.$minor.$patch_$YYYY-MM-DD` \
> Replace the `$major`, `$minor` and `$patch` with the new version and `$YYYY-MM-DD` with the current date in the format `YYYY-MM-DD`.
> 7. Move all changelog items from the `/changelog/unreleased/` folder into the `$major.$minor.$patch_$YYYY-MM-DD` folder created in previous step
> 8. Run `make l10n-pull` to pull the translations \
> This command will pull the latest translations from Transifex and save them in `/packages/*/l10n/` folders.
> 9. Run `make l10n-write` to write the translations \
> This command will use translations pulled in previous step and write them to the `/packages/*/l10n/translations.json` files.
> 10. Run `./dev/scripts/bump_versions.sh $major.$minor.$patch`. This script will bump the `package.json` files in all relevant packages and the sonar cloud project version \
> Replace the `$major`, `$minor` and `$patch` with the new version.
> 11. Commit and push your changes.
> 12. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-version-$major.$minor.$patch` branch into the new `stable-$major` branch
> 13. After merging, checkout the new `stable-$major` branch again
> 14. Run `./dev/scripts/create_and_push_tags.sh` \
> This script will create and push tags for the main app as well as all packages that need to be released
> 15. The GitHub release will be created automatically together with the release artifacts

Reviewer: ```suggestion
1. Checkout the latest `stable-$major` branch\
This branch will be used as the base for the new major version branch.\
Replace `$major` with the **last** released major version.
2. Create a new branch `stable-$major`.\
Replace `$major` with the **new** major version.
3. Push the newly created `stable-$major` branch
4. Create and checkout a new branch `chore/bump-version-$major.$minor.$patch` using the branch from previous step as base\
Replace `$major`, `$minor` and `$patch` with the new version.
5. Cherry pick all required changes from `master` into the current branch
6. Create a new sub-folder in the `/changelog/` folder using the release version and current date `mkdir $major.$minor.$patch_$YYYY-MM-DD`\
Replace `$major`, `$minor` and `$patch` with the new version and `$YYYY-MM-DD` with the current date in the format `YYYY-MM-DD`.
7. Move all changelog items from the `/changelog/unreleased/` folder into the `$major.$minor.$patch_$YYYY-MM-DD` folder created in the previous step
8. Run `make l10n-pull` to pull the translations\
This command will pull the latest translations from Transifex into the current branch and saves them in the `/packages/*/l10n/` folders.
9. Run `make l10n-write` to write the translations\
This command will use translations pulled in the previous step and writes them to the `/packages/*/l10n/translations.json` files.
10. Run `./dev/scripts/bump_versions.sh $major.$minor.$patch`.\
This script will bump the `package.json` files in all relevant packages and the sonar cloud project version\
Replace `$major`, `$minor` and `$patch` with the new version.
11. Commit and push your changes.
12. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-version-$major.$minor.$patch` branch into the new `stable-$major` branch
13. After merging, checkout the new `stable-$major` branch again
14. Run `./dev/scripts/create_and_push_tags.sh`\
This script will create and push tags for the main app as well as all packages that need to be released
15. The GitHub release will be created automatically together with the release artifacts
```

> #### Releasing new minor or patch version

Reviewer: ```suggestion
#### Releasing a New Minor or Patch Version
```

> 1. Checkout the latest `stable-$major` branch \
> This branch will be used as the base for the bump version branch. \
> Replace the `$major` with the last released major version.
> 2. Create a new branch `chore/bump-version-$major.$minor.$patch` \
> Replace the `$major`, `$minor` and `$patch` with the new version.
> 3. Cherry pick all required changes from `master` into the current branch
> 4. Create a new folder in `/changelog/` folder using the release version and current date `mkdir $major.$minor.$patch_$YYYY-MM-DD` \
> Replace the `$major`, `$minor` and `$patch` with the new version and `$YYYY-MM-DD` with the current date in the format `YYYY-MM-DD`.
> 5. Move all changelog items from the `/changelog/unreleased/` folder into the `$major.$minor.$patch_$YYYY-MM-DD` folder created in previous step
> 6. Run `make l10n-pull` to pull the translations \
> This command will pull the latest translations from Transifex and save them in `/packages/*/l10n/` folders.
> 7. Run `make l10n-write` to write the translations \
> This command will use translations pulled in previous step and write them to the `/packages/*/l10n/translations.json` files.
> 8. Run `./dev/scripts/bump_versions.sh $major.$minor.$patch`. This script will bump the `package.json` files in all relevant packages and the sonar cloud project version \
> Replace the `$major`, `$minor` and `$patch` with the new version.
> 9. Commit and push your changes.
> 10. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-version-$major.$minor.$patch` branch into the latest `stable-$major` branch
> 11. After merging, checkout the latest `stable-$major` branch again
> 12. Run `./dev/scripts/create_and_push_tags.sh` \
> This script will create and push tags for the main app as well as all packages that need to be released
> 13. The GitHub release will be created automatically together with the release artifacts

Reviewer: ```suggestion
1. Checkout the latest `stable-$major` branch\
This branch will be used as the base for the bump version branch.\
Replace the `$major` with the last released major version.
2. Create a new branch `chore/bump-version-$major.$minor.$patch`\
Replace `$major`, `$minor` and `$patch` with the new version.
3. Cherry pick all required changes from `master` into the current branch
4. Create a new sub-folder in `/changelog/` folder using the release version and current date `mkdir $major.$minor.$patch_$YYYY-MM-DD`\
Replace `$major`, `$minor` and `$patch` with the new version and `$YYYY-MM-DD` with the current date in the format `YYYY-MM-DD`.
5. Move all changelog items from the `/changelog/unreleased/` folder into the `$major.$minor.$patch_$YYYY-MM-DD` folder created in previous step
6. Run `make l10n-pull` to pull the translations\
This command will pull the latest translations from Transifex into the current branch and saves them in the `/packages/*/l10n/` folders.
7. Run `make l10n-write` to write the translations\
This command will use translations pulled in previous step and writes them to the `/packages/*/l10n/translations.json` files.
8. Run `./dev/scripts/bump_versions.sh $major.$minor.$patch`. This script will bump the `package.json` files in all relevant packages and the sonar cloud project version\
Replace the `$major`, `$minor` and `$patch` with the new version.
9. Commit and push your changes.
10. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-version-$major.$minor.$patch` branch into the latest `stable-$major` branch
11. After merging, checkout the latest `stable-$major` branch again
12. Run `./dev/scripts/create_and_push_tags.sh`\
This script will create and push tags for the main app as well as all packages that need to be released
13. The GitHub release will be created automatically together with the release artifacts
```

> #### Backporting bug/security fixes into older versions

Reviewer: ```suggestion
#### Backporting Bug/Security Fixes Into Older Versions
```

> 1. Checkout the target `stable-$major` branch \
> This branch will be used as the base for the new minor version branch. \
> Replace the `$major` with the target released major version.
> 2. Create a new branch `stable-$major.$minor`. \
> Replace the `$major` and `$minor` with the new version.
> 3. Push the newly created `stable-$major.$minor` branch
> 4. Create a new branch `chore/bump-version-$major.$minor.$patch` using the branch from previous step as base \
> Replace the `$major`, `$minor` and `$patch` with the new version.
> 5. Cherry pick all required changes from `master` into the current branch
> 6. Create a new folder in `/changelog/` folder using the release version and current date `mkdir $major.$minor.$patch_$YYYY-MM-DD` \
> Replace the `$major`, `$minor` and `$patch` with the new version and `$YYYY-MM-DD` with the current date in the format `YYYY-MM-DD`.
> 7. Move all changelog items from the `/changelog/unreleased/` folder into the `$major.$minor.$patch_$YYYY-MM-DD` folder created in previous step
> 8. Run `make l10n-pull` to pull the translations \
> This command will pull the latest translations from Transifex and save them in `/packages/*/l10n/` folders.
> 9. Run `make l10n-write` to write the translations \
> This command will use translations pulled in previous step and write them to the `/packages/*/l10n/translations.json` files.
> 10. Run `./dev/scripts/bump_versions.sh $major.$minor.$patch`. This script will bump the `package.json` files in all relevant packages and the sonar cloud project version \
> Replace the `$major`, `$minor` and `$patch` with the new version.
> 11. Commit and push your changes.
> 12. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-version-$major.$minor.$patch` branch into the new `stable-$major.$minor` branch
> 13. After merging, checkout the new `stable-$major.$minor` branch again
> 14. Run `./dev/scripts/create_and_push_tags.sh` \
> This script will create and push tags for the main app as well as all packages that need to be released
> 15. The GitHub release will be created automatically together with the release artifacts

Reviewer: ```suggestion
1. Checkout the target `stable-$major` branch\
This branch will be used as the base for the new minor version branch.\
Replace `$major` with the target released major version.
2. Create a new branch `stable-$major.$minor`.\
Replace the `$major` and `$minor` with the new version.
3. Push the newly created `stable-$major.$minor` branch
4. Create a new branch `chore/bump-version-$major.$minor.$patch` using the branch from previous step as base\
Replace the `$major`, `$minor` and `$patch` with the new version.
5. Cherry pick all required changes from `master` into the current branch
6. Create a new sub-folder in the `/changelog/` folder using the release version and current date `mkdir $major.$minor.$patch_$YYYY-MM-DD`\
Replace the `$major`, `$minor` and `$patch` with the new version and `$YYYY-MM-DD` with the current date in the format `YYYY-MM-DD`.
7. Move all changelog items from the `/changelog/unreleased/` folder into the `$major.$minor.$patch_$YYYY-MM-DD` folder created in previous step
8. Run `make l10n-pull` to pull the translations\
This command will pull the latest translations from Transifex into the current branch and saves them in the `/packages/*/l10n/` folders.
9. Run `make l10n-write` to write the translations\
This command will use translations pulled in previous step and writes them to the `/packages/*/l10n/translations.json` files.
10. Run `./dev/scripts/bump_versions.sh $major.$minor.$patch`.\
This script will bump the `package.json` files in all relevant packages and the sonar cloud project version\
Replace `$major`, `$minor` and `$patch` with the new version.
11. Commit and push your changes.
12. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-version-$major.$minor.$patch` branch into the new `stable-$major.$minor` branch
13. After merging, checkout the new `stable-$major.$minor` branch again
14. Run `./dev/scripts/create_and_push_tags.sh`\
This script will create and push tags for the main app as well as all packages that need to be released
15. The GitHub release will be created automatically together with the release artifacts
```

### Update Web in oCIS

> 1. In the [oCIS repository](https://github.com/owncloud/ocis), checkout master branch
> 2. Create a new branch `chore/bump-web-to-$major.$minor.$patch` \
> Replace the `$major`, `$minor` and `$patch` with the released Web version.
> 3. In `.drone.env`, update `WEB_COMMITID` to the commit id of the Web release tag
> 4. In `.drone.env`, update `WEB_BRANCH` to the Web release stable branch
> 5. In `services/web/Makefile`, update `WEB_ASSETS_VERSION` to the Web release version
> 6. Add a new `enhancement` changelog item into the `/changelog/unreleased/` folder with the changelog from the Web release and link to the GitHub release
> 7. Commit and push your changes
> 8. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-web-to-$major.$minor.$patch` branch into the master branch

Reviewer: ```suggestion
1. In the [oCIS repository](https://github.com/owncloud/ocis), checkout the `master` branch
2. Create a new branch `chore/bump-web-to-$major.$minor.$patch`\
Replace `$major`, `$minor` and `$patch` with the released Web version.
3. In `.drone.env`, update `WEB_COMMITID` to the commit id of the Web release tag
4. In `.drone.env`, update `WEB_BRANCH` to the Web release stable branch
5. In `services/web/Makefile`, update `WEB_ASSETS_VERSION` to the Web release version
6. Add a new `enhancement` changelog item into the `/changelog/unreleased/` folder with the changelog from the Web release and link to the GitHub release
7. Commit and push your changes
8. Create a PR with tag `[full-ci]` in the title at the beginning to merge the `chore/bump-web-to-$major.$minor.$patch` branch into the master branch
```
